vktrace/vktrace_replay/vkreplay_vkreplay.cpp

```diff
@@ -2256,16 +2112,182 @@ VkResult vkReplay::manually_replay_vkAllocateMemory(packet_vkAllocateMemory *pPa
+                    *((VkDeviceMemory *)((PBYTE)pPacket->header + pPacket->header->size - sizeof(VkDeviceMemory *)));
+
+                // Save the index where we will start the next search for vkAM
+                amSearchPos = amIdx + 1;
+                break;
+            }
+        }
+        if (amIdx == portabilityTable.size()) {
+            // Didn't find the current vkAM packet, something is wrong with the trace file.
+            // Just use the index from the trace file and attempt to continue.
+            vktrace_LogError("Replay of vkAllocateMemory() failed, trace file may be corrupt.");
             replayResult = m_vkFuncs.real_vkAllocateMemory(remappedDevice, pPacket->pAllocateInfo, NULL, &local_mem.replayGpuMem);
-        } else {
-            vktrace_LogError("vkAllocateMemory() failed, couldn't find memory type for memoryTypeIndex");
-            return VK_ERROR_VALIDATION_FAILED_EXT;
+            fseek(tracefp, saveFilePos, SEEK_SET);
+            goto wrapItUp;
+        }
+
+        // Search forward from amIdx for all vkBIM and vkBBM calls that bind this memory.
+        // For each one we find, search backwards for the vkGIMR/vkGBMR call that uses the
+        // same image/buffer. If we don't find one, generate an error and do the best we can.
+        // Sum up the size returned by the vkGIMR/vkGBMR calls - we'll use that size for the replay vkAM.
+        // Find the greatest alignment in the vkGIMR/vkGBMR calls - we'll use that alignment for the replay vkAM.
+        foundBindMem = false;
+        for (size_t i = amIdx + 1; i < portabilityTable.size(); i++) {
+            FSEEK(tracefp, (long)portabilityTable[i], SEEK_SET);
+            FREAD(&packetHeader1, sizeof(vktrace_trace_packet_header), 1, tracefp);  // Read the packet header
+
+            if (packetHeader1.packet_id == VKTRACE_TPI_VK_vkBindImageMemory ||
+                packetHeader1.packet_id == VKTRACE_TPI_VK_vkBindBufferMemory) {
+                assert(packetHeader1.size == sizeof(packetHeader1) + sizeof(bimPacket));
+                FREAD(&bimPacket, sizeof(bimPacket), 1, tracefp);
+            }
+
+            if (packetHeader1.packet_id == VKTRACE_TPI_VK_vkFreeMemory) {
+                FREAD(&freeMemoryPacket, sizeof(freeMemoryPacket), 1, tracefp);
+                if (freeMemoryPacket.memory == traceAllocateMemoryRval)
+                    // Found a free of this memory, end the forward search
+                    break;
+            }
+
+            if ((packetHeader1.packet_id == VKTRACE_TPI_VK_vkBindImageMemory ||
+                 packetHeader1.packet_id == VKTRACE_TPI_VK_vkBindBufferMemory) &&
+                traceAllocateMemoryRval == bimPacket.memory) {
+                // A vkBIM/vkBBM binds memory allocated by this vkAM call.
+                foundBindMem = true;
+
+                // Search backwards for the vkGIMR/vkGBMR call.
+                int64_t j;
+                for (j = amIdx - 1; j >= 0; j--) {
```

Comment:
This for compares int64_t to size_t, (signed vs unsigned). Be better define j uint64_t